framework/decode/dx12_replay_consumer_base.cpp

```diff
@@ -3771,5 +3801,577 @@ std::wstring Dx12ReplayConsumerBase::ConstructObjectName(format::HandleId captur
     return constructed_name;
 }
 
+void Dx12ReplayConsumerBase::PreCall_ID3D12GraphicsCommandList_ResourceBarrier(
+    const ApiCallInfo&                                    call_info,
+    DxObjectInfo*                                         object_info,
+    UINT                                                  NumBarriers,
+    StructPointerDecoder<Decoded_D3D12_RESOURCE_BARRIER>* pBarriers)
+{
+    auto barriers = pBarriers->GetMetaStructPointer();
+    for (uint32_t i = 0; i < NumBarriers; ++i)
+    {
+        auto resource_id          = barriers[i].Transition->pResource;
+        auto resource_object_info = GetObjectInfo(resource_id);
+        auto resource_extra_info  = GetExtraInfo<D3D12ResourceInfo>(resource_object_info);
+
+        if (options_.dump_resources_type != DumpResourcesType::kNone)
+        {
+            // This should be in the override functionsince it modifies the parameters. But here is ok.
+            barriers[i].Transition->decoded_value->StateBefore = resource_extra_info->current_state;
+            if (barriers[i].Transition->decoded_value->StateAfter & D3D12_RESOURCE_STATE_ALL_SHADER_RESOURCE)
+            {
+                // shader resources can't change state after SetGraphicsRootDescriptorTable.
+                // Adding a COPY_SOURCE state here, so it doesn't need to change state for copy resources.
+                barriers[i].Transition->decoded_value->StateAfter |= D3D12_RESOURCE_STATE_COPY_SOURCE;
```

Comment:
shader resources can't change state after SetGraphicsRootDescriptorTable. In this case, adding a COPY_SOURCE state for shader resource, so it doesn't need to change state for copy resources.